x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/DataNodeRequestSender.java

```diff
@@ -204,7 +203,8 @@ private void trySendingRequestsForPendingShards(TargetShards targetShards, Compu
                         }
                     }
                     for (ShardId shardId : pendingShardIds) {
-                        if (targetShards.getShard(shardId).remainingNodes.isEmpty()) {
+                        if (targetShards.getShard(shardId).remainingNodes.isEmpty()
+                            && (isRetryableFailure(shardFailures.get(shardId)) == false || pendingRetries.contains(shardId))) {
```

Comment:
It was previously possible that between executing line 195 and line 206 one of the data nodes returns with NoSuchShard exception and add a new `pendingShardId`. As a result such shard might still not have `remainingNodes` nodes to query. This change supposed to detect such situations and delay such resolution to the next round.

This was detected by `testSearchWhileRelocating` integration test that is fairly slow and expensive so I added `testRetryMultipleMovedShards` unit test that made it much easier to reproduce.